CleverSDK/Classes/CLVOAuthManager.m

```diff
@@ -136,39 +213,69 @@ + (BOOL)handleURL:(NSURL *)url sourceApplication:(NSString *)sourceApplication a
     NSDictionary *parameters = @{@"code": code, @"grant_type": @"authorization_code", @"redirect_uri": [self redirectUri]};
 
     [tokens POST:@"oauth/tokens" parameters:parameters progress:nil success:^(NSURLSessionDataTask *task, id responseObject) {
+        CLVOAuthManager *manager = [self sharedManager];
+        manager.alreadyMissedCode = NO;
         // verify that the client id is what we expect
         if ([responseObject objectForKey:@"access_token"]) {
             [self setAccessToken:responseObject[@"access_token"]];
-            [[NSNotificationCenter defaultCenter] postNotificationName:CLVAccessTokenReceivedNotification object:self];
+            [self callSucessHandler];
         } else {
             // if no access token was received, consider this a failure
-            [[NSNotificationCenter defaultCenter] postNotificationName:CLVOAuthAuthorizeFailedNotification object:self];
+            [self callFailureHandler];
         }
     } failure:^(NSURLSessionDataTask *task, NSError *error) {
         CLVOAuthManager *manager = [self sharedManager];
+        manager.alreadyMissedCode = NO;
         manager.errorMessage = [NSString stringWithFormat:@"%@",  [error localizedDescription]];
-        [[NSNotificationCenter defaultCenter] postNotificationName:CLVOAuthAuthorizeFailedNotification object:self];
+        [self callFailureHandler];
     }];
     return YES;
 }
 
-+ (void)successHandler:(void (^)(NSString *accessToken))successHandler failureHandler:(void (^)(NSString *errorMessage))failureHandler {
-    CLVOAuthManager *manager = [self sharedManager];
-    manager.successHandler = successHandler;
-    manager.failureHandler = failureHandler;
-}
-
 + (void)callSucessHandler {
     CLVOAuthManager *manager = [self sharedManager];
+    // iOS 8 - call success handler
+    // iOS 9/10 - dismiss SFSafariViewController before calling success handler
+    // iOS 11+ - call success handler
+    if (SYSTEM_VERSION_LESS_THAN(@"11.0") && SYSTEM_VERSION_GREATER_THAN_OR_EQUAL_TO(@"9.0")) {
+        // Must dismiss SFSafariViewController
+        [manager.uiDelegate dismissViewControllerAnimated:NO completion:^{
+            if (manager.successHandler) {
+                manager.successHandler(self.accessToken);
+            } else if (manager.delegate) {
+                [manager.delegate signInToClever:[self accessToken] withError:nil];
+            }
+        }];
+        return;
+    }
     if (manager.successHandler) {
-        manager.successHandler([self accessToken]);
+        manager.successHandler(self.accessToken);
+    } else if (manager.delegate) {
+        [manager.delegate signInToClever:[self accessToken] withError:nil];
     }
 }
 
 + (void)callFailureHandler {
+    [self logout];
```

Comment:
Can you explain what scenario the logout on failure covers?

Reply:
If the SDK already has an existing `accessToken` and a new login fails, I think it would be a better experience for the user not to have access to the old accessToken. Seems like a confusing developer experience to me. Don't have a strong opinion on this issue though - happy to remove it if you think we should.

Reply:
No, I just didn't know what scenario it covered. Think it's fine to keep it in